File: native/ios/AngularMobile/Sources/ViewRegistry.swift
import Foundation
import UIKit

/// View Registry
///
/// Manages the registry of all native views created by Angular Platform Mobile.
/// Provides thread-safe access to views by their IDs.
public class ViewRegistry {

    // MARK: - Properties

    private var views: [String: UIView] = [:]
    private var viewIdCounter: Int = 0
    private weak var rootContainer: UIView?
    private let lock = NSLock()

    // MARK: - Public API

    /// Generate a unique view ID
    public func generateViewId() -> String {
        lock.lock()
        defer { lock.unlock() }

        viewIdCounter += 1
        return "view_\(viewIdCounter)"
    }

    /// Register a view with its ID
    public func register(viewId: String, view: UIView) {
        lock.lock()
        defer { lock.unlock() }

        views[viewId] = view
        view.accessibilityIdentifier = viewId
    }

    /// Get a view by ID
    public func get(viewId: String) -> UIView? {
        lock.lock()
        defer { lock.unlock() }

        return views[viewId]
    }

    /// Check if a view exists
    public func has(viewId: String) -> Bool {
        lock.lock()
        defer { lock.unlock() }

        return views[viewId] != nil
    }

    /// Unregister a view and remove from parent
    public func unregister(viewId: String) {
        lock.lock()
        let view = views.removeValue(forKey: viewId)
        lock.unlock()

        DispatchQueue.main.async {
            view?.removeFromSuperview()
        }
    }

    /// Set the root container for the view hierarchy
    public func setRootContainer(_ container: UIView) {
        lock.lock()
        defer { lock.unlock() }

        rootContainer = container
    }

    /// Get the root container
    public func getRootContainer() -> UIView? {
        lock.lock()
        defer { lock.unlock() }

        return rootContainer
    }

    /// Get all registered view IDs
    public func getAllViewIds() -> [String] {
        lock.lock()
        defer { lock.unlock() }

        return Array(views.keys)
    }

    /// Get the total count of registered views
    public func getViewCount() -> Int {
        lock.lock()
        defer { lock.unlock() }

        return views.count
    }

    /// Clear all views
    public func clear() {
        lock.lock()
        let allViews = Array(views.values)
        views.removeAll()
        lock.unlock()

        DispatchQueue.main.async {
            allViews.forEach { $0.removeFromSuperview() }
        }
    }

    /// Find view ID by view instance
    public func findViewId(for view: UIView) -> String? {
        lock.lock()
        defer { lock.unlock() }

        if let identifier = view.accessibilityIdentifier, views[identifier] === view {
            return identifier
        }

        return views.first { $0.value === view }?.key
    }
}

File: native/android/runtime/src/main/kotlin/com/pegasusheavy/angularmobile/ViewRegistry.kt
package com.pegasusheavy.angularmobile

import android.view.View
import android.view.ViewGroup
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger

/**
 * View Registry
 *
 * Manages the registry of all native views created by Angular Platform Mobile.
 * Provides thread-safe access to views by their IDs.
 */
class ViewRegistry {
    private val views = ConcurrentHashMap<String, View>()
    private val viewIdCounter = AtomicInteger(0)
    private var rootContainer: ViewGroup? = null

    /**
     * Generate a unique view ID
     */
    fun generateViewId(): String {
        return "view_${viewIdCounter.incrementAndGet()}"
    }

    /**
     * Register a view with its ID
     */
    fun register(viewId: String, view: View) {
        views[viewId] = view
        view.tag = viewId
    }

    /**
     * Get a view by ID
     */
    fun get(viewId: String): View? {
        return views[viewId]
    }

    /**
     * Check if a view exists
     */
    fun has(viewId: String): Boolean {
        return views.containsKey(viewId)
    }

    /**
     * Unregister a view and remove from parent
     */
    fun unregister(viewId: String) {
        val view = views.remove(viewId)
        view?.let {
            (it.parent as? ViewGroup)?.removeView(it)
        }
    }

    /**
     * Set the root container for the view hierarchy
     */
    fun setRootContainer(container: ViewGroup) {
        rootContainer = container
    }

    /**
     * Get the root container
     */
    fun getRootContainer(): ViewGroup? {
        return rootContainer
    }

    /**
     * Get all registered view IDs
     */
    fun getAllViewIds(): Set<String> {
        return views.keys.toSet()
    }

    /**
     * Get the total count of registered views
     */
    fun getViewCount(): Int {
        return views.size
    }

    /**
     * Clear all views
     */
    fun clear() {
        views.values.forEach { view ->
            (view.parent as? ViewGroup)?.removeView(view)
        }
        views.clear()
    }

    /**
     * Find view ID by view instance
     */
    fun findViewId(view: View): String? {
        return view.tag as? String ?: views.entries.find { it.value === view }?.key
    }
}
